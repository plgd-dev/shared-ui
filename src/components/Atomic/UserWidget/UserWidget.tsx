import { useRef, useState } from 'react'
import classNames from 'classnames'
import { useIntl } from 'react-intl'
import { useClickOutside } from '../../../common/hooks'
import { messages as t } from './UserWidget.i18n'
import { Props } from './UserWidget.types'
import './UserWidget.scss'
import Avatar from 'react-avatar'
import { useAuth } from 'oidc-react'

const UserWidget = (props: Props) => {
    const { logout } = props
    const [expanded, setExpand] = useState(false)
    const { isLoading, userData } = useAuth()
    const { formatMessage: _ } = useIntl()
    const clickRef = useRef<HTMLDivElement>(null)
    useClickOutside(clickRef, () => setExpand(false))

    if (isLoading || !userData) {
        return null
    }

    return (
        <div className='status-bar-item' id='user-widget' ref={clickRef}>
            <div className='toggle' onClick={() => setExpand(!expanded)}>
                <div className='user-avatar'>
                    {userData.profile.picture ? (
                        <img alt={userData.profile.name} src={userData.profile.picture} />
                    ) : (
                        <Avatar color='#255897' name={userData.profile.name} round='50%' size='30' />
                    )}
                </div>
                {userData.profile.name}
                <i
                    className={classNames('fas', {
                        'fa-chevron-down': !expanded,
                        'fa-chevron-up': expanded,
                    })}
                />
            </div>
            {expanded && (
                <div className='content'>
                    <span onClick={() => logout()}>{_(t.logOut)}</span>
                </div>
            )}
        </div>
    )
}

UserWidget.displayName = 'UserWidget'

export default UserWidget
