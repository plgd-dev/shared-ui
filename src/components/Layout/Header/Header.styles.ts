import { css } from '@emotion/react'
import { colors } from '../../Atomic/_utils/colors'
import { ThemeType, getThemeColor } from '../../Atomic/_theme'

export const header = (theme: ThemeType) => css`
    background-color: ${getThemeColor(theme, `Header.background`)};
    height: 85px;
    padding: 20px 40px;
    box-sizing: border-box;
    display: flex;
    align-items: center;
    justify-content: space-between;
    border-bottom: 1px solid ${getThemeColor(theme, `Header.borderBottom`)};
`

export const left = css`
    display: flex;
    align-items: center;
`

export const iconCollapse = css`
    color: ${colors.neutral500};
`

export const breadcrumbs = css`
    padding-left: 35px;
`

export const right = css`
    display: flex;
`

export const actions = css`
    list-style: none;
    padding: 0;
    margin: 0;
    display: flex;
`

export const action = css`
    display: flex;
    justify-content: center;
    align-items: center;
    color: ${colors.neutral500};
`
