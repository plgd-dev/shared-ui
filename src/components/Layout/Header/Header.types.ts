import { ReactNode } from 'react'

export type Props = {
    breadcrumbs?: ReactNode
    configButton?: ReactNode
    notificationCenter?: ReactNode
    userWidget?: ReactNode
}
