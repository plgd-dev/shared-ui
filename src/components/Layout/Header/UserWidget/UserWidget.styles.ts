import { css } from '@emotion/react'
import { colors } from '../../../Atomic/_utils/colors'
import { ThemeType, get } from '../../../Atomic/_theme'

export const widgetReference = css`
    display: flex;
    align-items: stretch;
`

export const userWidget = css`
    padding-left: 12px;
    display: flex;
`

export const clickable = css`
    cursor: pointer;
`

export const widgetArrow = css`
    padding-left: 12px;
    display: flex;
    align-items: center;
    cursor: pointer;
`

export const icon = css`
    transition: all 0.3s;
`

export const iconActive = css`
    transform: rotate(180deg);
`

export const name = (theme: ThemeType) => css`
    font-family: 'Poppins', sans-serif;
    font-style: normal;
    font-weight: 500;
    font-size: 16px;
    line-height: 160%;
    color: ${get(theme, `UserWidget.name.color`)};
`

export const description = css`
    font-family: 'Poppins', sans-serif;
    font-style: normal;
    font-weight: 400;
    font-size: 12px;
    line-height: 18px;
    color: ${colors.neutral500};
    opacity: 0.8;
`

export const image = (theme: ThemeType) => css`
    width: 44px;
    height: 44px;
    flex: 0 0 44px;
    margin-right: 12px;

    img {
        border-radius: 50%;
        width: 100%;
        height: auto;
    }

    .sb-avatar__text {
        background: ${get(theme, `UserWidget.image.background`)} !important;
    }
`

export const logoutIcon = (theme: ThemeType) => css`
    display: flex;
    align-self: center;
    margin-left: 36px;

    color: ${get(theme, `UserWidget.logout.color`)};
    cursor: pointer;
    transition: all 0.3s;

    &:hover {
        color: ${get(theme, `UserWidget.logout.hover.color`)};
    }
`

export const logoutIconBtn = (theme: ThemeType) => css`
    color: ${get(theme, `UserWidget.logout.color`)};
    cursor: pointer;
    transition: all 0.3s;

    &:hover {
        color: ${get(theme, `UserWidget.logout.hover.color`)};
    }
`

export const floatingMenu = (theme: ThemeType) => css`
    z-index: 10;
    background: ${get(theme, `UserWidget.floatingMenu.background`)};
    border: 1px solid ${get(theme, `UserWidget.floatingMenu.border`)};
    box-shadow: 0 30px 40px rgba(28, 52, 99, 0.1);
    border-radius: 8px;
    padding: 10px 20px;
`

export const item = (theme: ThemeType) => css`
    display: flex;
    align-items: center;
    text-decoration: none;
    color: ${colors.neutral500};
    transition: all 0.3s;
    margin: 6px 0;
    padding: 4px 0;
    cursor: pointer;

    &:hover {
        color: ${get(theme, `UserWidget.floatingMenu.item.hover.color`)};
    }
`
