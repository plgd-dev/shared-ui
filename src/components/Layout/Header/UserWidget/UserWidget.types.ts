import { ReactNode } from 'react'

export type DropdownItemType = {
    title: string
    onClick: () => void
    dataTestId?: string
}

export type Props = {
    defaultOpen?: boolean
    description?: string
    dropdownItems?: DropdownItemType[]
    image?: ReactNode
    loading?: boolean
    name: string
    dataTestId?: string
}
