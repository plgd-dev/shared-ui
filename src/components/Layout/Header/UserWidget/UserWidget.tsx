import { FC, useRef, useState } from 'react'
import { Props } from './UserWidget.types'
import * as styles from './UserWidget.styles'
import Avatar from 'react-avatar'
import { useClickOutside } from '../../../../common/hooks'
import { offset, shift, useFloating } from '@floating-ui/react'
import Icon from '../../../Atomic/Icon'

const UserWidget: FC<Props> = (props) => {
    const { name, description, defaultOpen, dropdownItems, image, loading } = props
    const [open, setOpen] = useState(defaultOpen ?? false)
    const clickRef = useRef<HTMLDivElement>(null)
    useClickOutside(clickRef, () => setOpen(false))

    const { x, y, reference, floating, strategy } = useFloating({
        placement: 'bottom-end',
        strategy: 'fixed',
        middleware: [shift(), offset(4)],
    })

    if (loading) {
        return null
    }

    const handleClick = () => {
        dropdownItems && setOpen(!open)
    }

    return (
        <div ref={clickRef}>
            <div css={styles.widgetReference} onClick={handleClick} ref={reference}>
                <div css={[styles.userWidget, dropdownItems && styles.clickable]}>
                    {image && <div css={styles.image}>{typeof image === 'string' ? <img alt={name} src={image} /> : image}</div>}
                    {!image && (
                        <div css={styles.image}>
                            <Avatar color='#255897' name={name} round='50%' size='44' />
                        </div>
                    )}
                    <div>
                        <div css={styles.name}>{name}</div>
                        {description && <div css={styles.description}>{description}</div>}
                    </div>
                </div>
                {dropdownItems && (
                    <div css={styles.widgetArrow}>
                        <Icon css={[styles.icon, open && styles.iconActive]} icon='arrow-down' size={12} />
                    </div>
                )}
            </div>

            {dropdownItems && open && (
                <div
                    css={styles.floatingMenu}
                    ref={floating}
                    style={{
                        position: strategy,
                        top: y ?? 0,
                        left: x ?? 0,
                        width: 'max-content',
                    }}
                >
                    {dropdownItems.map((item) => (
                        <div css={styles.item} onClick={item.onClick} key={item.title}>
                            {item.title}
                        </div>
                    ))}
                </div>
            )}
        </div>
    )
}

UserWidget.displayName = 'UserWidget'

export default UserWidget
