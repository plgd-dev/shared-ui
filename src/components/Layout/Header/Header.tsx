import { FC } from 'react'
import { Props } from './Header.types'
import * as styles from './Header.styles'
// import { convertSize, IconSettings } from '../../Atomic/Icon'

const Header: FC<Props> = (props) => {
    const { breadcrumbs, configButton, notificationCenter, userWidget } = props

    return (
        <div css={styles.header}>
            <div css={styles.left}>
                <div id='header-icon-collapse-portal-target'></div>
                {breadcrumbs && <div css={styles.breadcrumbs}>{breadcrumbs}</div>}
            </div>
            <div css={styles.right}>
                <ul css={styles.actions}>
                    {configButton && <li>{configButton}</li>}
                    {/* <li>*/}
                    {/*    <a css={styles.actionItem} href='#'>*/}
                    {/*        <IconSettings {...convertSize(20)} />*/}
                    {/*    </a>*/}
                    {/* </li>*/}
                    {notificationCenter && <li>{notificationCenter}</li>}
                </ul>
                {userWidget}
            </div>
        </div>
    )
}

Header.displayName = 'Header'

export default Header
