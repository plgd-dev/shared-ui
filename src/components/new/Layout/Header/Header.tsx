import { FC, SyntheticEvent } from "react";
import { Props } from './Header.types'
import * as styles from './Header.styles'
import { Icon } from '../../Icon'
import isFunction from 'lodash/isFunction'

const Header: FC<Props> = (props) => {
    const { breadcrumbs, onCollapseToggle, userWidget } = props

    const handleCollapseToggle = (e: SyntheticEvent) => {
        e.preventDefault()
        e.stopPropagation()

        isFunction(onCollapseToggle) && onCollapseToggle()
    }

    return (
        <div css={styles.header}>
            <div css={styles.left}>
                <a css={styles.iconCollapse} href='#' onClick={(e) => handleCollapseToggle(e)}>
                    <Icon icon='collapse' size={24} />
                </a>
                {breadcrumbs && <div css={styles.breadcrumbs}>{breadcrumbs}</div>}
            </div>
            <div css={styles.right}>
                <ul css={styles.actions}>
                    <li>
                        <a css={styles.actionItem} href='#'>
                            <Icon icon='settings' size={20} />
                        </a>
                    </li>
                    <li>
                        <a css={styles.actionItem} href='#'>
                            <Icon icon='notification' size={20} />
                        </a>
                    </li>
                </ul>
                {userWidget}
            </div>
        </div>
    )
}

Header.displayName = 'Header'

export default Header
