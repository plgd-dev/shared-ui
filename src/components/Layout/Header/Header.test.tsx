import React from 'react'
import { render } from '@testing-library/react'
import Header from './Header'
import UserWidget from './UserWidget'

describe('<Header>', () => {
    it('render correctly - snapshot', () => {
        const userData = {
            profile: {
                name: '',
                family_name: '',
                picture: '',
            },
        }
        const { asFragment } = render(
            <Header
                breadcrumbs={<div id='breadcrumbsPortalTarget'></div>}
                userWidget={<UserWidget description={userData?.profile?.family_name} image={userData?.profile?.picture} name={userData?.profile?.name || ''} />}
            />
        )

        expect(asFragment()).toMatchSnapshot()
    })
})
