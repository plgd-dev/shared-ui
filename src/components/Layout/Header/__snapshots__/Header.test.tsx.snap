// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header> render correctly - snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  background-color: #fff;
  height: 85px;
  padding: 20px 40px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  border-bottom: 1px solid #e6e9ed;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  padding-left: 35px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  list-style: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-6 {
  padding-left: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-7 {
  width: 44px;
  height: 44px;
  -webkit-flex: 0 0 44px;
  -ms-flex: 0 0 44px;
  flex: 0 0 44px;
  margin-right: 12px;
}

.emotion-7 img {
  border-radius: 50%;
  width: 100%;
  height: auto;
}

.emotion-8 {
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 500;
  font-size: 16px;
  line-height: 160%;
  color: #0a2965;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        id="header-icon-collapse-portal-target"
      />
      <div
        class="emotion-2"
      >
        <div
          id="breadcrumbsPortalTarget"
        />
      </div>
    </div>
    <div
      class="emotion-3"
    >
      <ul
        class="emotion-4"
      />
      <div>
        <div
          class="emotion-5"
        >
          <div
            class="emotion-6"
          >
            <div
              class="emotion-7"
            >
              <div
                class=" sb-avatar"
                style="display: inline-block; vertical-align: middle; width: 44px; height: 44px; border-radius: 50%; font-family: Helvetica, Arial, sans-serif;"
              >
                <div
                  class=" sb-avatar__text"
                  style="width: 44px; height: 44px; line-height: initial; text-align: center; color: rgb(255, 255, 255); background: rgb(37, 88, 151); border-radius: 50%;"
                >
                  <div
                    style="display: table; table-layout: fixed; width: 100%; height: 100%;"
                  >
                    <span
                      style="display: table-cell; vertical-align: middle; font-size: 100%; white-space: nowrap;"
                    >
                      <span />
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                class="emotion-8"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
